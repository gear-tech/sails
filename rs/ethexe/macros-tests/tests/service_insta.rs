use proc_macro2::TokenStream;
use quote::quote;
use sails_macros_core::__gservice_internal as gservice;

#[test]
fn works_with_basics() {
    let input = quote! {
        impl SomeService {
            #[export]
            pub async fn do_this(&mut self, p1: u32, p2: String) -> u32 {
                p1
            }

            #[export]
            pub fn this(&self, p1: bool) -> bool {
                p1
            }
        }
    };

    let result = gservice(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_lifetimes_and_generics() {
    let input = quote! {
        impl<'a, 'b, T, U> SomeService<'a, 'b, T, U>
        where T : Clone,
              U: Iterator<Item = u32> {

            #[export]
            pub fn do_this(&mut self) -> u32 {
                42
            }
        }
    };

    let result = gservice(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_extends() {
    let args = quote! {
        extends = [ExtendedService1, ExtendedService2],
        //arg42 = "Hello, World!"
    };
    let input = quote! {
        impl SomeService {
            #[export]
            pub fn do_this(&mut self) -> u32 {
                42
            }
        }
    };

    let result = gservice(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_events() {
    let args = quote! {
        events = MyEvents,
    };
    let input = quote! {
        impl MyServiceWithEvents {
            #[export]
            pub fn do_this(&mut self) -> u32 {
                42
            }

            #[export]
            pub fn this(&self) -> bool {
                true
            }
        }
    };

    let result = gservice(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_lifetimes_and_events() {
    let args = quote! {
        events = MyEvents,
    };
    let input = quote! {
        impl<'a, T> MyGenericEventsService<'a, T>
        where
            T: Clone,
        {
            #[export]
            pub fn do_this(&mut self) -> u32 {
                42
            }
        }
    };

    let result = gservice(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_extends_and_lifetimes() {
    let args = quote! {
        extends = [BaseWithLifetime<'a>]
    };
    let input = quote! {
        impl<'a> ExtendedWithLifetime<'a> {
            #[export]
            pub fn extended_name(&self) -> String {
                "extended-name".to_string()
            }

            #[export]
            pub fn name(&self) -> String {
                "extended".to_string()
            }
        }
    };

    let result = gservice(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_methods_with_lifetimes() {
    let input = quote! {
        impl ReferenceService {
            #[export]
            pub fn baked(&self) -> &'static str {
                "Static str!"
            }

            #[export]
            pub fn incr(&mut self) -> &'static ReferenceCount {
                unsafe {
                    COUNTER.0 += 1;
                    &*ptr::addr_of!(COUNTER)
                }
            }

            #[export]
            pub fn add_byte(&mut self, byte: u8) -> &'static [u8] {
                unsafe {
                    BYTES.push(byte);
                    &*ptr::addr_of!(BYTES)
                }
            }

            #[export]
            pub async fn first_byte<'a>(&self) -> Option<&'a u8> {
                unsafe { BYTES.first() }
            }

            #[export]
            pub async fn last_byte<'a>(&self) -> Option<&'a u8> {
                unsafe { BYTES.last() }
            }
        }
    };

    let result = gservice(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_crate_path() {
    let args = quote!(crate = sails_rename,);
    let input = quote! {
        impl SomeService {
            #[export]
            pub async fn do_this(&mut self, p1: u32, p2: String) -> u32 {
                p1
            }

            #[export]
            pub fn this(&self, p1: bool) -> bool {
                p1
            }
        }
    };

    let result = gservice(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_reply_with_value() {
    let input = quote! {
        impl SomeService {
            #[export]
            pub async fn do_this(&mut self, p1: u32, p2: String) -> CommandReply<u32> {
                (p1, 100_000_000_000).into()
            }

            #[export]
            pub fn this(&self, p1: bool) -> bool {
                p1
            }
        }
    };

    let result = gservice(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_allow_attrs() {
    let input = quote! {
        #[allow(warnings)]
        #[allow(clippy::all)]
        impl SomeService {
            #[allow(unused)]
            #[export]
            pub async fn do_this(&mut self, p1: u32, p2: String) -> u32 {
                p1
            }

            #[export]
            pub fn this(&self, p1: bool) -> bool {
                p1
            }
        }
    };

    let result = gservice(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_docs() {
    let input = quote! {
        impl SomeService {
            /// `DoThis` command
            /// Second line
            #[export]
            pub async fn do_this(&mut self, p1: u32, p2: String) -> u32 {
                p1
            }

            /// `This` query
            #[export]
            pub fn this(&self, p1: bool) -> bool {
                p1
            }
        }
    };

    let result = gservice(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_special_lifetimes_and_events() {
    let args = quote! {
        events = MyEvents,
    };
    let input = quote! {
        impl<T> MyGenericEventsService<'_, '_, T>
        where
            T: Clone,
        {
            #[export]
            pub fn do_this(&mut self) -> u32 {
                42
            }
        }
    };

    let result = gservice(args, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn works_with_export() {
    let input = quote! {
        impl SomeService {
            #[export(route = "DoSomething", unwrap_result)]
            pub async fn do_this(&mut self, p1: u32, p2: String) -> Result<(u32, String), String> {
                Ok((p1, p2))
            }

            #[export]
            pub fn this(&self, p1: bool) -> bool {
                p1
            }
        }
    };

    let result = gservice(TokenStream::new(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}
