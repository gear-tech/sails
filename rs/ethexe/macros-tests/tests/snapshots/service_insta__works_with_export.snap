---
source: macros-tests/tests/service_insta.rs
expression: result
---
impl SomeService {
    #[export(route = "DoSomething", unwrap_result)]
    pub async fn do_this(
        &mut self,
        p1: u32,
        p2: String,
    ) -> Result<(u32, String), String> {
        Ok((p1, p2))
    }
    pub fn this(&self, p1: bool) -> bool {
        p1
    }
}
pub struct SomeServiceExposure<T: sails_rs::gstd::services::Service> {
    message_id: sails_rs::MessageId,
    route: &'static [u8],
    #[cfg(not(target_arch = "wasm32"))]
    inner: sails_rs::Box<T>,
    #[cfg(not(target_arch = "wasm32"))]
    inner_ptr: *const T,
    #[cfg(target_arch = "wasm32")]
    inner: T,
    base: T::BaseExposures,
}
impl<T: sails_rs::gstd::services::Service> sails_rs::gstd::services::Exposure
for SomeServiceExposure<T> {
    fn message_id(&self) -> sails_rs::MessageId {
        self.message_id
    }
    fn route(&self) -> &'static [u8] {
        self.route
    }
}
impl SomeServiceExposure<SomeService> {
    pub async fn do_this(
        &mut self,
        p1: u32,
        p2: String,
    ) -> Result<(u32, String), String> {
        let exposure_scope = sails_rs::gstd::services::ExposureCallScope::new(self);
        self.inner.do_this(p1, p2).await
    }
    pub fn this(&self, p1: bool) -> bool {
        let exposure_scope = sails_rs::gstd::services::ExposureCallScope::new(self);
        self.inner.this(p1)
    }
    pub fn try_handle(
        &mut self,
        input: &[u8],
        result_handler: fn(&[u8], u128),
    ) -> Option<()> {
        use sails_rs::gstd::InvocationIo;
        use sails_rs::gstd::services::Exposure;
        if let Ok(request) = some_service_meta::__ThisParams::decode_params(input) {
            let result = self.this(request.p1);
            let value = 0u128;
            some_service_meta::__ThisParams::with_optimized_encode(
                &result,
                self.route().as_ref(),
                |encoded_result| result_handler(encoded_result, value),
            );
            return Some(());
        }
        None
    }
    pub async fn try_handle_async(
        &mut self,
        input: &[u8],
        result_handler: fn(&[u8], u128),
    ) -> Option<()> {
        use sails_rs::gstd::InvocationIo;
        use sails_rs::gstd::services::Exposure;
        if let Ok(request) = some_service_meta::__DoSomethingParams::decode_params(
            input,
        ) {
            let result = self.do_this(request.p1, request.p2).await.unwrap();
            let value = 0u128;
            some_service_meta::__DoSomethingParams::with_optimized_encode(
                &result,
                self.route().as_ref(),
                |encoded_result| result_handler(encoded_result, value),
            );
            return Some(());
        }
        None
    }
    pub fn try_handle_solidity(
        &mut self,
        method: &[u8],
        input: &[u8],
    ) -> Option<(sails_rs::Vec<u8>, u128, bool)> {
        if method == &[16u8, 84u8, 104u8, 105u8, 115u8] {
            let (_, _encode_reply, p1): (u128, bool, bool) = sails_rs::alloy_sol_types::SolValue::abi_decode_params(
                    input,
                    false,
                )
                .ok()?;
            let result = self.this(p1);
            let value = 0u128;
            let output = if _encode_reply {
                let message_id = sails_rs::alloy_primitives::B256::new(
                    self.message_id.into_bytes(),
                );
                sails_rs::alloy_sol_types::SolValue::abi_encode_sequence(
                    &(message_id, result),
                )
            } else {
                sails_rs::alloy_sol_types::SolValue::abi_encode_sequence(&(result,))
            };
            return Some((output, value, _encode_reply));
        }
        None
    }
    pub async fn try_handle_solidity_async(
        &mut self,
        method: &[u8],
        input: &[u8],
    ) -> Option<(sails_rs::Vec<u8>, u128, bool)> {
        if method
            == &[
                44u8, 68u8, 111u8, 83u8, 111u8, 109u8, 101u8, 116u8, 104u8, 105u8, 110u8,
                103u8,
            ]
        {
            let (_, _encode_reply, p1, p2): (u128, bool, u32, String) = sails_rs::alloy_sol_types::SolValue::abi_decode_params(
                    input,
                    false,
                )
                .ok()?;
            let result = self.do_this(p1, p2).await.unwrap();
            let value = 0u128;
            let output = if _encode_reply {
                let message_id = sails_rs::alloy_primitives::B256::new(
                    self.message_id.into_bytes(),
                );
                sails_rs::alloy_sol_types::SolValue::abi_encode_sequence(
                    &(message_id, result),
                )
            } else {
                sails_rs::alloy_sol_types::SolValue::abi_encode_sequence(&(result,))
            };
            return Some((output, value, _encode_reply));
        }
        None
    }
    pub fn check_asyncness(&self, input: &[u8]) -> Option<bool> {
        use sails_rs::gstd::InvocationIo;
        if !<SomeService as sails_rs::meta::ServiceMeta>::ASYNC {
            return Some(false);
        }
        if let Ok(is_async) = some_service_meta::__DoSomethingParams::check_asyncness(
            input,
        ) {
            return Some(is_async);
        }
        if let Ok(is_async) = some_service_meta::__ThisParams::check_asyncness(input) {
            return Some(is_async);
        }
        None
    }
}
impl sails_rs::gstd::services::Service for SomeService {
    type Exposure = SomeServiceExposure<Self>;
    type BaseExposures = ();
    fn expose(
        self,
        message_id: sails_rs::MessageId,
        route: &'static [u8],
    ) -> Self::Exposure {
        #[cfg(not(target_arch = "wasm32"))]
        let inner_box = sails_rs::Box::new(self);
        #[cfg(not(target_arch = "wasm32"))]
        let inner = inner_box.as_ref();
        #[cfg(target_arch = "wasm32")]
        let inner = &self;
        Self::Exposure {
            message_id,
            route,
            base: (),
            #[cfg(not(target_arch = "wasm32"))]
            inner_ptr: inner_box.as_ref() as *const Self,
            #[cfg(not(target_arch = "wasm32"))]
            inner: inner_box,
            #[cfg(target_arch = "wasm32")]
            inner: self,
        }
    }
}
impl sails_rs::meta::ServiceMeta for SomeService {
    type CommandsMeta = some_service_meta::CommandsMeta;
    type QueriesMeta = some_service_meta::QueriesMeta;
    type EventsMeta = some_service_meta::EventsMeta;
    const BASE_SERVICES: &'static [sails_rs::meta::AnyServiceMetaFn] = &[];
    const ASYNC: bool = true;
}
mod some_service_meta {
    use super::*;
    use sails_rs::{Decode, TypeInfo};
    use sails_rs::gstd::InvocationIo;
    #[derive(Decode, TypeInfo)]
    #[codec(crate = sails_rs::scale_codec)]
    #[scale_info(crate = sails_rs::scale_info)]
    pub struct __DoSomethingParams {
        pub(super) p1: u32,
        pub(super) p2: String,
    }
    impl InvocationIo for __DoSomethingParams {
        const ROUTE: &'static [u8] = &[
            44u8, 68u8, 111u8, 83u8, 111u8, 109u8, 101u8, 116u8, 104u8, 105u8, 110u8,
            103u8,
        ];
        type Params = Self;
        const ASYNC: bool = true;
    }
    #[derive(Decode, TypeInfo)]
    #[codec(crate = sails_rs::scale_codec)]
    #[scale_info(crate = sails_rs::scale_info)]
    pub struct __ThisParams {
        pub(super) p1: bool,
    }
    impl InvocationIo for __ThisParams {
        const ROUTE: &'static [u8] = &[16u8, 84u8, 104u8, 105u8, 115u8];
        type Params = Self;
        const ASYNC: bool = false;
    }
    #[derive(TypeInfo)]
    #[scale_info(crate = sails_rs::scale_info)]
    pub enum CommandsMeta {
        DoSomething(__DoSomethingParams, (u32, String)),
    }
    #[derive(TypeInfo)]
    #[scale_info(crate = sails_rs::scale_info)]
    pub enum QueriesMeta {
        This(__ThisParams, bool),
    }
    #[derive(TypeInfo)]
    #[scale_info(crate = sails_rs::scale_info)]
    pub enum NoEvents {}
    pub type EventsMeta = NoEvents;
}
impl sails_rs::solidity::ServiceSignature for SomeService {
    const METHODS: &'static [sails_rs::solidity::MethodExpo] = &[
        (
            &[
                44u8, 68u8, 111u8, 83u8, 111u8, 109u8, 101u8, 116u8, 104u8, 105u8, 110u8,
                103u8,
            ] as &[u8],
            "DoSomething",
            <<(
                u128,
                bool,
                u32,
                String,
            ) as sails_rs::alloy_sol_types::SolValue>::SolType as sails_rs::alloy_sol_types::SolType>::SOL_NAME,
            <<(
                sails_rs::alloy_primitives::B256,
                (u32, String),
            ) as sails_rs::alloy_sol_types::SolValue>::SolType as sails_rs::alloy_sol_types::SolType>::SOL_NAME,
        ),
        (
            &[16u8, 84u8, 104u8, 105u8, 115u8] as &[u8],
            "This",
            <<(
                u128,
                bool,
                bool,
            ) as sails_rs::alloy_sol_types::SolValue>::SolType as sails_rs::alloy_sol_types::SolType>::SOL_NAME,
            <<(
                sails_rs::alloy_primitives::B256,
                bool,
            ) as sails_rs::alloy_sol_types::SolValue>::SolType as sails_rs::alloy_sol_types::SolType>::SOL_NAME,
        ),
    ];
}
